spec/rubocop/cop/rspec/verified_double_reference_spec.rb

```diff
@@ -0,0 +1,41 @@
+# frozen_string_literal: true
+
+RSpec.describe RuboCop::Cop::RSpec::VerifiedDoubleReference do
+  context 'when EnforcedStyle is const' do
+    let(:cop_config) do
+      { 'EnforcedStyle' => 'const' }
+    end
+
+    it 'does not flag a violation when using a const reference' do
+      expect_no_offenses('instance_double(ClassName)')
+    end
+
+    it 'flags a violation when using a string reference' do
+      expect_offense(<<-RUBY)
+        instance_double("ClassName")
```

Comment:
Unrelated to this line specifically. What if an instance variable is used?
```ruby
instance_double(class_name)
```

Reply:
@pirj It will flag a violation. My thought is instance variables to classes could be used by overriding the cop when necessary. Open to opinions. I am making assumptions about how common instance variable class references are used in tests anecdotally.

An alternative would be to allow `lvar` nodes for one or both styles

Reply:
I'm on the fence with this.
Could you please run the cop against https://github.com/pirj/real-world-rspec to see if there are legitimate use cases that would trigger the cop?

Reply:
From real-world-rspec, here are a few cases I found that result in errors (from `Hash#fetch`) when running this new cop:

- https://github.com/rspec/rspec-mocks/blob/7823b0b542456cfd6193293e3f19e01b7eb934ee/spec/rspec/mocks/formatting_spec.rb#L80-L92
- https://github.com/rspec/rspec-mocks/blob/7823b0b542456cfd6193293e3f19e01b7eb934ee/spec/rspec/mocks/spy_spec.rb#L66-L68
- https://github.com/rspec/rspec-mocks/blob/7823b0b542456cfd6193293e3f19e01b7eb934ee/spec/rspec/mocks/verifying_doubles/construction_spec.rb#L61
- https://github.com/diaspora/diaspora/blob/5d81555ae1009fd35ef1f14473d8cf0956d91d6f/spec/integration/federation/receive_federation_messages_spec.rb#L115
- https://github.com/thoughtbot/administrate/blob/dcc137567b75580a21401558dec3ce0294b794fc/spec/administrate/views/fields/date_time/_index_spec.rb#L6-L12
- https://github.com/instructure/canvas-lms/blob/050120c81d3c6ee362e10d0244f4780f488ecd0d/spec/controllers/gradebooks_controller_spec.rb#L3133

Reply:
@bquorning Do you mind sharing the setup you used to run the spec against the real-world-rspec? I was about to dig into issues with `bundle exec rubocop -c ./config/default.yml --only RSpec/VerifiedDoubleReference ../real-world-rspec`

It's now my opinion that this cop should allow local and instance variables by default. This could be taken further to allow anything that isn't the opposite style of the current EnforcedStyle. This should handle the `(array)` from https://github.com/rspec/rspec-mocks/blob/7823b0b542456cfd6193293e3f19e01b7eb934ee/spec/rspec/mocks/formatting_spec.rb#L80-L92.

An alternative could be to add a configuration option for both instance and local variables that determines whether they are allowed. These options would default to true.

Reply:
> Do you mind sharing the setup you used to run the spec against the real-world-rspec? I was about to dig into issues with `bundle exec rubocop -c ./config/default.yml --only RSpec/VerifiedDoubleReference ../real-world-rspec`

I did almost that, but found that the cloned projects’ rubocop configs were causing various problems, so I removed them first.
```
rm -v ../real-world-rspec/**/.rubocop.yml
bundle exec rubocop --only RSpec/VerifiedDoubleReference -- ../real-world-rspec/
```

Reply:
> should allow any node type that isn't the opposite style of the current EnforcedStyle

:+1:

Reply:
@prij @bquorning updated to now allow any node type that isn't the opposite style of the current EnforcedStyle

When running against real-world-rspec there are no cop errors, and get this nice output:
> 56602 files inspected, 529 offenses detected, 529 offenses auto-correctable